File: 31.NextPermutation/main.cpp
#include<iostream>
#include<vector>


using std::vector, std::cout, std::endl, std::string, std::max;


class Solution {
 public:
  void nextPermutation(vector<int> nums) {
    if (nums.size() == 1)
      return;
    int right = nums.size() - 1;
    int left = right - 1;
    while (left >= 0 && nums[left] >= nums[right]) {
      left--;
      right--;
    }
    if (left >= 0) {
      int right = nums.size() - 1;
      while (nums[right] <= nums[left])
        right--;
      iter_swap(nums.begin() + right, nums.begin() + left);
    }
    reverse(nums.begin() + left + 1, nums.end());
  }
};


int main() {
  vector<int> input = {0, 2, 5, 8, 17};
  for (auto s : input)
    cout << s << ' ';
  cout << '\n';
  Solution().nextPermutation(input);
  for (auto s : input)
    cout << s << ' ';
  return 0;
}

File: 31.NextPermutation/main.py
class Solution:
    def nextPermutation(self, nums):
        i = len(nums) - 1
        while i > 0:
            if nums[i - 1] < nums[i]:
                break
            i = i - 1
        i = i - 1
        j = len(nums) - 1
        while j > i:
            if nums[j] > nums[i]:
                break
            j = j - 1
        nums[i], nums[j] = nums[j], nums[i]  
        nums[i + 1:] = sorted(nums[i + 1:])
